_posts/2021-01-11-batch-fine-grained-fault-tolerance.md

```diff
@@ -0,0 +1,117 @@
+2. **Execution time with frequent failures**: The results “in the middle”, with failure frequencies of 6, 5, 4, 3 and 2 minutes show that the pipelined mode execution gets unfeasible at some point: If failures happen on average every 3 minutes, the average execution time reaches more than 60 minutes, for failures every 2 minutes the time spikes to more than 120 minutes. The pipelined job is unable to finish the execution, only if we happen to find a window where no failure is induced for 15 minutes, the pipelined job manages to produce the final result. For more frequent failures, the pipelined mode did not manage to finish at all.
+3. **How many failures can the batch mode sustain?** The last numbers, with failure frequencies between 60 and 20 seconds are probably a bit unrealistic for real world scenarios. But we wanted to investigate how frequent failures can become for the batch mode to become unfeasible. With failures induced every 30 seconds, the average execution time is 30 minutes. In other words, even if you have two failures per minute, your execution time only doubles in this case. The batch mode is much more predictable and well behaved when it comes to execution times.
+
+
+## Conclusion {#conclusion}
+
+Based on these results, it makes a lot of sense to use the batch execution mode for batch workloads, as the resource consumption and overall execution times are substantially lower compared to the pipelined execution mode.
+
+In general, we recommend conducting your own performance experiments on your own hardware and with your own workloads, as results might vary from what we’ve presented here. Despite the findings here, the pipelined mode probably has some performance advantages in environments with rare failures and slower I/O (for example when using spinning disks, or network attached disks). On the other hand, CPU intensive workloads might benefit from the batch mode even in slow I/O environments.
+
+We should also note that the caching (and subsequent reprocessing on failure) only works if the cached results are still present -- this is currently only the case, if the TaskManager survives a failure. However, this is an unrealistic assumption as many failures would cause the TaskManager process to die. To mitigate this limitation, data processing frameworks employ external shuffle services that persist the cached results independent of the data processing framework. Since Flink 1.9, there is support for a [pluggable shuffle service](https://cwiki.apache.org/confluence/display/FLINK/FLIP-31%3A+Pluggable+Shuffle+Service), and there are tickets for adding implementations for [YARN](https://issues.apache.org/jira/browse/FLINK-13247) and [Kubernetes](https://issues.apache.org/jira/browse/FLINK-13246). Once these implementations are added, TaskManagers can recover cached results even if the process or machine got killed.
```

Comment:
```suggestion
We should also note that the caching (and subsequent reprocessing on failure) only works if the cached results are still present -- this is currently only the case, if the TaskManager survives a failure. However, this is an unrealistic assumption as many failures would cause the TaskManager process to die. To mitigate this limitation, data processing frameworks employ external shuffle services that persist the cached results independent of the data processing framework. Since Flink 1.9, there is support for a [pluggable shuffle service](https://cwiki.apache.org/confluence/display/FLINK/FLIP-31%3A+Pluggable+Shuffle+Service), and there are tickets for adding implementations for YARN ([FLINK-13247](https://issues.apache.org/jira/browse/FLINK-13247)) and Kubernetes ([FLINK-13246](https://issues.apache.org/jira/browse/FLINK-13246)). Once these implementations are added, TaskManagers can recover cached results even if the process or machine got killed.
```
Minor thing: I added the FLINK issues in brackets to make it more explicit.